### PR TITLE
Backend and business logic

### PR DESCRIPTION
**Pull request by RichardDev101**
First pull request for backend and business logic with integration and unit tests included.

### CODE REVIEW

File: backend/src/main/java/dev/projects/backend/collection/Advertisement.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    private AdvertisementStatus advertisementStatus;
+    private PaymentCategory paymentCategory;//payment Categories which the vendor pays to us
+    private Company company;
+    private List<BusinessCategory>businessCategories;
+    private List<String> photosID;
+    private String title;
+    private String aboutYourself; // Section where one writes about himself in a shot manner
+    private String detailInformationForService; //Description about the Service someone ist offering
+    private float averagePrice; // Price someone requests for his service
+    private List<PriceCategory> priceCategories; //per hour/ per day... etc.
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/collection/Advertisement.java:39]: no need for comments in production

File: backend/src/main/java/dev/projects/backend/controller/AdController.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api/ad")
+public class AdController {
+
+    @Autowired
+    private AdService adService;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/AdController.java:18]: thats one way to use DI. but i recommend changing to constrctor injecting. 
change to : 
`private final AdService adService`
in combination with the @RequiredArgsConstructor Annotation

File: backend/src/main/java/dev/projects/backend/controller/AdController.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    //CREATE
+    @PostMapping()
+    //@PreAuthorize("hasAuthority('LoginRole=ADMIN')")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/AdController.java:22]: no comment in production

File: backend/src/main/java/dev/projects/backend/controller/AdController.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                .detailInformationForService(ad.getDetailInformationForService())
+                .averagePrice(ad.getAveragePrice())
+                .priceCategories(ad.getPriceCategories())
+                .contacts(ad.getContacts())
+                .locations(ad.getLocations())
+                .personsID(ad.getPersonsID())
+                .build();
+        return adService.updateAd(updateAd, id);
+    }
+    //DELETE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/AdController.java:84]: no comment in production

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import dev.projects.backend.service.PersonService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/api")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:15]: every endpoint is starting with 'person' u should use it up here at the global level

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+public class PersonController {
+
+    @Autowired
+    private PersonService personService;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:19]: constructor injection?

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        PersonDTO masterAdmin = new PersonDTO();
+        masterAdmin.setUserName("${ENV_UN_Admin}");
+        masterAdmin.setPassWord("${ENV_PW_Admin}");
+        masterAdmin.setLoginRole(LoginRole.ADMIN);
+        saveAdmin(masterAdmin);
+        return ResponseEntity.status(HttpStatus.CREATED).body("MasterAdmin has been successfully created.");
+    }
+    @PostMapping("/person/admin")
+    //@PreAuthorize("hasAuthority('LoginRole=ADMIN')")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:40]: no comment in production

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                .lastName(person.getLastName())
+                .contactDetails(person.getContactDetails())
+                .addresses(person.getAddresses())
+                .loginRole(person.getLoginRole())
+                .gpsData(person.getGpsData())
+                .build();
+        return personService.saveAdmin(addAdmin);
+    }
+    @PostMapping("/person/editor")
+    //@PreAuthorize("hasAuthority('LoginRole=ADMIN')")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:55]: same

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                .lastName(person.getLastName())
+                .contactDetails(person.getContactDetails())
+                .addresses(person.getAddresses())
+                .loginRole(person.getLoginRole())
+                .gpsData(person.getGpsData())
+                .build();
+        return personService.saveEditor(addEditor);
+    }
+    @PostMapping("/person/user")
+    //@PreAuthorize("hasAuthority('LoginRole=ADMIN')")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:70]: same

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                .lastName(person.getLastName())
+                .contactDetails(person.getContactDetails())
+                .addresses(person.getAddresses())
+                .loginRole(person.getLoginRole())
+                .gpsData(person.getGpsData())
+                .build();
+        return personService.saveUser(addUser);
+    }
+
+    //READ
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:85]: same

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    public List<Person> getPersonsByRole(@RequestParam("role") LoginRole role){
+        return personService.getPersonsByRole(role);
+    }
+
+    @GetMapping("/person/{id}")
+    public Person getPersonById(@PathVariable String id){
+        return personService.getPersonById(id);
+    }
+
+    //UPDATE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:100]: same

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                .firstName(person.getFirstName())
+                .lastName(person.getLastName())
+                .contactDetails(person.getContactDetails())
+                .addresses(person.getAddresses())
+                .loginRole(person.getLoginRole())
+                .gpsData(person.getGpsData())
+                .build();
+        return personService.update(updatePerson, id);
+    }
+    //DELETE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:116]: same

File: backend/src/main/java/dev/projects/backend/controller/PhotoController.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+
+@RestController
+@RequestMapping("/api/photo")
+public class PhotoController {
+
+    @Autowired
+    private PhotoService photoService;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PhotoController.java:22]: constructor injection

File: backend/src/main/java/dev/projects/backend/controller/VendorController.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+@RestController
+@RequestMapping("/api")
+public class VendorController {
+
+
+    @Autowired
+    private VendorService vendorService;
+
+    //CREATE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/VendorController.java:19]: no comments

File: backend/src/main/java/dev/projects/backend/controller/VendorController.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    @PostMapping("/vendor")
+    //@PreAuthorize("hasAuthority('LoginRole=ADMIN')")
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/VendorController.java:21]: same

File: backend/src/main/java/dev/projects/backend/controller/VendorController.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    public Vendor saveVendor(@RequestBody VendorDTO vendor){
+            Vendor newVendor = new Vendor(vendor);
+        return vendorService.saveVendor(newVendor);
+    }
+    //READ
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/VendorController.java:26]: same

File: backend/src/main/java/dev/projects/backend/controller/VendorController.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+    @GetMapping("/vendors")
+    public List<Vendor> getAllVendors(){
+        return vendorService.getAllVendors();
+    }
+    @GetMapping("/vendor/{id}")
+    public Vendor getVendorWithId(@PathVariable String id){
+        return vendorService.getVendorById(id);
+    }
+    //UPDATE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/VendorController.java:35]: same

File: backend/src/main/java/dev/projects/backend/controller/VendorController.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    @PutMapping("/vendor/{id}")
+    public Vendor update(@RequestBody VendorDTO vendor, @PathVariable String id){
+        Vendor newVendor = new Vendor(vendor);
+        newVendor.setPersonId(id);
+        return vendorService.updateVendor(newVendor, id);
+    }
+    //DELETE
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/VendorController.java:42]: same

File: backend/src/main/java/dev/projects/backend/dto/AdvertisementDTO.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    private AdvertisementStatus advertisementStatus;
+    private PaymentCategory paymentCategory;//payment Categories which the vendor pays to us
+    private Company company;
+    private List<BusinessCategory> businessCategories;
+    private List<String> photosID;
+    private String title;
+    private String aboutYourself; // Section where one writes about himself in a shot manner
+    private String detailInformationForService; //Description about the Service someone ist offering
+    private float averagePrice; // Price someone requests for his service
+    private List<PriceCategory> priceCategories; //per hour/ per day... etc.
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/dto/AdvertisementDTO.java:27]: no comment

File: backend/src/main/java/dev/projects/backend/service/AdServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Optional;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Service
+public class AdServiceImpl implements AdService{
+
+    @Autowired
+    private AdRepository adRepository;
+    @Autowired
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/service/AdServiceImpl.java:23]: probably better to use constructor injection
and are interfaces really needed?

File: backend/src/main/java/dev/projects/backend/service/PersonServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Service
+public class PersonServiceImpl implements PersonService {
+
+    @Autowired
+    private PersonRepository personRepository;
+    @Autowired
+    private GenerateUUIDService uuid;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/service/PersonServiceImpl.java:23]: same

File: backend/src/main/java/dev/projects/backend/service/PhotoServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Service
+public class PhotoServiceImpl implements PhotoService{
+
+    @Autowired
+    private PhotoRepository photoRepository;
+    @Autowired
+    private GenerateUUIDService uuid;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/service/PhotoServiceImpl.java:25]: same

File: backend/src/main/java/dev/projects/backend/service/VendorServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Service
+public class VendorServiceImpl implements VendorService{
+
+    @Autowired
+    private VendorRepository vendorRepository;
+    @Autowired
+    private GenerateUUIDService uuid;
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/service/VendorServiceImpl.java:23]: same

File: backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+    @Autowired
+    MockMvc mockMvc;
+
+    @Test
+    @DirtiesContext
+    @WithMockUser
+    void testCheckAndCreateMasterAdmin() throws Exception {
+        mockMvc.perform(MockMvcRequestBuilders.post("/api/person/master-admin")
+                        .with(csrf()))
+                .andExpect(status().is(201));
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java:34]: u are expecting "MasterAdmin has been successfully created" as well

File: backend/src/test/java/dev/projects/backend/controller/PhotoControllerTest.java [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+
+        when(photoService.addPhoto(anyString(), any(MockMultipartFile.class).thenReturn("photoId123"));
+
+        mockMvc.perform(multipart("/api/photo")
+                        .file(file))
+                .andExpect(status().isOk())
+                .andExpect(content().string("photoId123"));
+
+        verify(photoService).addPhoto(anyString(), any(MockMultipartFile.class));
+    }*/
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/controller/PhotoControllerTest.java:49]: no comment

File: backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java [hunk 26]
```diff
@@ -0,0 +1,310 @@
...
+                        {
+                                "userName": "${ENV_UN_Admin}",
+                                "passWord": "${ENV_PW_Admin}",
+                                "loginRole": "ADMIN"
+                              }
+                             ]
+                        """))
+                .andExpect(jsonPath("$[0].personId").isNotEmpty());
+    }
+
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java:50]: missing test for "MasterAdmin already exists." ?

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+                if(personList.get(i).getUserName().equals("${ENV_UN_Admin}")){
+                    return ResponseEntity.ok("MasterAdmin already exists.");
+                }
+            }
+        }
+        PersonDTO masterAdmin = new PersonDTO();
+        masterAdmin.setUserName("${ENV_UN_Admin}");
+        masterAdmin.setPassWord("${ENV_PW_Admin}");
+        masterAdmin.setLoginRole(LoginRole.ADMIN);
+        saveAdmin(masterAdmin);
```
Comment (Kreyno93) [inline backend/src/main/java/dev/projects/backend/controller/PersonController.java:36]: why calling endpoint method? either implement it in "checkAndCreateMasterAdmin" or use a Service

File: backend/src/test/java/dev/projects/backend/service/VendorServiceImplTest.java [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.*;
+
+@SpringBootTest
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/service/VendorServiceImplTest.java:18]: no need in unit tests

File: backend/src/test/java/dev/projects/backend/service/PhotoServiceImplTest.java [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.io.IOException;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.*;
+
+@SpringBootTest
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/service/PhotoServiceImplTest.java:18]: same

File: backend/src/test/java/dev/projects/backend/service/PersonServiceImplTest.java [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+import dev.projects.backend.enums.LoginRole;
+import dev.projects.backend.repository.PersonRepository;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.*;
+
```
Comment (Kreyno93) [inline backend/src/test/java/dev/projects/backend/service/PersonServiceImplTest.java:13]: u did it here

**Review by Kreyno93 — CHANGES_REQUESTED** [review]

**Review by Kreyno93 — CHANGES_REQUESTED** [review]
u are not really using your services. all the logic is in the controller. either use only controllers. or shift the logic to the services

**Review by Kreyno93 — CHANGES_REQUESTED** [review]
Interfaces arent really needed. and i would change the order of the naming, if interfaces arent really needed

File: backend/src/main/java/dev/projects/backend/collection/Advertisement.java (hunk 1 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/collection/Advertisement.java:39]: gelöscht

File: backend/src/main/java/dev/projects/backend/controller/AdController.java (hunk 2 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/controller/AdController.java:18]: @RequiredArgsConstructor with private final AdService adService; has been implemented

File: backend/src/main/java/dev/projects/backend/controller/AdController.java (hunk 3 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/controller/AdController.java:22]: deleted

File: backend/src/main/java/dev/projects/backend/controller/AdController.java (hunk 4 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/controller/AdController.java:84]: deleted

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java (hunk 5 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/controller/PersonController.java:15]: I'll will make later an distinction between person and persons when the result will be a list. Therefore, I'll keep it like it is.

File: backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java (hunk 26 above)
Comment (RichardDev101) [reply backend/src/test/java/dev/projects/backend/controller/PersonControllerTest.java:50]: position has been deleted

File: backend/src/main/java/dev/projects/backend/controller/PersonController.java (hunk 27 above)
Comment (RichardDev101) [reply backend/src/main/java/dev/projects/backend/controller/PersonController.java:36]: want to handle the master admin creation from the frontend if needed

**Comment by RichardDev101** [comment]
there were quit handy during the development process, but if the you dont not like them... deleted

**Pull request closed without merging**